### PR TITLE
InputControl: Reverting recent changes, which affected updates

### PR DESCRIPTION
**Pull request by ItsJonQ**
This update reverts 2 recent updates on the `InputControl` component:

https://github.com/WordPress/gutenberg/pull/25609
https://github.com/WordPress/gutenberg/pull/25753

I've noticed some issues in the BoxControl component in Gutenberg with the updates. In particular...

* adding new values and pressing `Tab` to commit
* drag to update no longer works :(

Screen Shot 2020-10-07 at 12 08 36 PM

cc'ing @stokesman 

Looking back at the initial issue (https://github.com/WordPress/gutenberg/issues/24460), these updates are attempting to fix some number input entering.

~I haven't been able to replicate the issues. Maybe I'm not entering things correctly (to break it) in my keyboard?~
~I'd love to know more info about this.~
~On top of reverts, I'd like to add solutions for the issues mentioned in https://github.com/WordPress/gutenberg/issues/24460~

## Updates

Beyond the revert, the changes can be seen here:

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R66-R83

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R30-R42

### 📹 Demo + Walkthrough

I recorded a video demo + walkthrough that showcases the issue and describes the fixes:
https://www.loom.com/share/22b6d3b1b61140e4a667ed61a6858f52

### ⚛️ Fixing State Handling

The issue was that the inner number input wasn't correctly handling and coordinating values/changes between the UI element (`input[type="number"]`) and the state that exists within the wrapping `RangeControl` component.

The solution involved swapping `useControlledState` for a simple `useState` + `useEffect` hook combination for state management and syncing within the inner number `InputField`.

The solution also involved (re)adding the HTML validation step before "commiting" (firing the `onChange`) callback. This allows for "invalid" state to be held within the number input before pushing it upwards.

What this means, is that users will be able to **type** in values like `123`, when the `min` attribute is set to `20`.

Previously, it would **jump** to `20` because typing in the initial character (`1`) would trigger onChange immediately... which clamps the value and propagates it updates.

This is no longer the case! (Thank goodness)

### 🗜 Clamping onChange

Secondly, the final value that is sent from the `RangeControl` onChange callback is clamped. Previously, this was not the case. We also didn't spot this problem due to the above mentioned issue.

## How has this been tested?

* Confirm reverting the commits fixes BoxControl issues and overall drag to update interaction
* Confirm `RangeControl` state flow is working

(Note: For drag to update, I think order matters for the `onMouseDown` callback. This isn't as tricky as the update that adjusts the update mechanism to rely on the `blur` event)

## How to test

* Run `npm run dev`
* Add a Cover block
* Adjust the height using the input control. This should work.
* Add a Columns block
* Interfacing with the RangeControl + accompanying number input should work.

## Checklist:
- [x] My code is tested.
- [n/a] My code follows the WordPress code style. 
- [n/a] My code follows the accessibility standards. 
- [n/a] My code has proper inline documentation. 
- [n/a] I've included developer documentation if appropriate. 
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by stokesman** [comment]
@ItsJonQ, I've taken a look at this and found fixes. That drag to update breakage is definitely something I should have caught. I remember wondering if that #25753 fix would interfere with dragging but obviously forgot to check 🤦 . The other is simpler to fix yet more mysterious. The `UnitControls` in `BoxControl` are all ending up with `isPressEnterToChange` as `true` and thus no `onChange` firing as input is made. Explicitly setting it to `false` works to fix it but I thought that was the default so I don't really understand 😕 .

I'll put it in a PR.

**Comment by ItsJonQ** [comment]
@stokesman Thank you! I'll take a look at RangeControl. I left a comment in your draft PR as well :)

**Comment by ItsJonQ** [comment]
Update: I believe the issue is isolated within `RangeControl` (as you noted @stokesman )

I recorded a GIF demo below:

range-control-input-issue

The inputs are using a min/max of `20` and `400`.
There are 3 "number" inputs:

1. HTML Native `input`
2. `NumberControl`
3. `NumberControl` within `RangeControl`

The HTML `input` and unwrapped `NumberControl` work as expected.
That is, entering the number `1` is accepted.

However... attempting to enter `1` in the `RangeControl` input jumps to `20`.

I may be wrong.. but I suspect this may have happened before. The [validation logic](https://github.com/WordPress/gutenberg/pull/23006/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6L75) existed before the refactor.

---

With all that said! Will see if I can come up with a solution.

If not, I think it's probably better to merge the revert as it'll fix the inputs in other areas

---

P.S.  @stokesman Just wanted to thank you for your care and patience with this one! It's tricky. I appreciate you diving in and helping 🙏

**Comment by ItsJonQ** [comment]
Update: @stokesman I think I got it!!

Will push update + record a demo very soon

**Comment by ItsJonQ** [comment]
Okay! I think this is ready.

Beyond the revert, the changes can be seen here:

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R66-R83

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R30-R42

## 📹 Demo + Walkthrough

I recorded a video demo + walkthrough that showcases the issue and describes the fixes:
https://www.loom.com/share/22b6d3b1b61140e4a667ed61a6858f52

## ⚛️ Fixing State Handling

The issue was that the inner number input wasn't correctly handling and coordinating values/changes between the UI element (`input[type="number"]`) and the state that exists within the wrapping `RangeControl` component.

The solution involved swapping `useControlledState` for a simple `useState` + `useEffect` hook combination for state management and syncing within the inner number `InputField`.

The solution also involved (re)adding the HTML validation step before "commiting" (firing the `onChange`) callback. This allows for "invalid" state to be held within the number input before pushing it upwards.

What this means, is that users will be able to **type** in values like `123`, when the `min` attribute is set to `20`.

Previously, it would **jump** to `20` because typing in the initial character (`1`) would trigger onChange immediately... which clamps the value and propagates it updates.

This is no longer the case! (Thank goodness)

## 🗜 Clamping onChange

Secondly, the final value that is sent from the `RangeControl` onChange callback is clamped. Previously, this was not the case. We also didn't spot this problem due to the above mentioned issue.

**Comment by stokesman** [comment]
I took this for a spin to look for a couple of tangential and obscure issues I thought may reappear. In doing so I ran across a new one which I'll describe first. In WordPress, while editing a spacer block, clear the number input, and then try typing any number besides '1'. It'll give you a '1' anyway. This seems to be unique to the Spacer block as I couldn't reproduce it in Storybook.

One of the obscure issues I mentioned is also with the Spacer block and can be reproduced by clearing the number input then dragging the in-canvas resize handle. The input won't update once the drag ends and after that direct input is not possible (the range input slider can recover it though).

<details>
 <summary>Screen recording of both of the above issues</summary>

</details>

It could be that both of those are related to the reset that can be triggered by `onChange` and the fact that the Spacer block doesn't handle it well. It errors with: "Warning: Received NaN for the `max` attribute. If this is expected, cast the value to a string." but script execution continues.
<details>
 <summary>See recording</summary>
 
</details>

There also seems to be some new issues with resets when clearing number inputs of range controls with the Gallery and Button blocks, but I'm going to take a break for now.

**Comment by ItsJonQ** [comment]
@stokesman Amazing! Thank you so much. This is really helpful 🙏

**Comment by ItsJonQ** [comment]
@stokesman Haii! I just pushed a fix. I think this does it 🤞 .
I've written inline comments to describe the flow.

https://github.com/WordPress/gutenberg/pull/25913/commits/68aa804d14d2fed95d4b6ee2846baee06c3321c5#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R72

**Comment by ItsJonQ** [comment]
**Unrelated update!**

This PR was tremendously helpful for me.

### 🔦 Spotlighting

I appreciated it as it highlights some of the trickiness that happens with seemingly simple input controls. It's also helped me rethink and refine approaches for improving how state could be handled within this controls (as a single unit), and how it could be shared (across multiple units, like a slider and an input).

### ✨ Updates
I've applied some of the ideas from this PR (and some of the explorations from @stokesman ) into the [`TextInput` component](https://github.com/ItsJonQ/g2/blob/master/packages/components/src/TextInput/useTextInput.js#L42) within the [G2 Components project](https://g2components.wordpress.com/about/).

Here's a quick gif demo that shows the use cases (mentioned above) working with a mixture of input interactions, and state flowing from internal and external changes:

Screen Capture on 2020-10-08 at 19-59-02

The most exciting thing (for me) is that how simple the UI code is in combining a `Slider` control and a `TextInput` control together. It looks like this:

```jsx
const CombinedControl = ({ max, min, onChange, value }) => {
	return (
		<Grid>
			<Slider max={max} min={min} onChange={onChange} value={value} />
			<TextInput
				max={max}
				min={min}
				onChange={onChange}
				type="number"
				value={value}
			/>
		</Grid>
	);
};
```

### ♻️ State Management
The biggest difference between how the UI works in the above example and the current `RangeControl` component.. is that the `Slider` and `TextInput` are **separate** components. They can be composed together with a simple wrapper.

Whereas the current `RangeControl` does this.. kind awkward wrangling of an `input[type="range"]` with an accompanying number input.

The strategy for controlling state flow is also different.

The Gutenberg `InputControl` relies on a reducer pattern.

The G2 `TextInput` uses a pubsub like pattern that's similar to that of [React Recoil](https://github.com/facebookexperimental/Recoil). This state management system is powered by [Zustand](https://github.com/pmndrs/zustand), which is used in libraries like `react-spring` (makes of Zustand!).

If you're curious, here's the state management for `TextInput` component:
https://github.com/ItsJonQ/g2/blob/master/packages/components/src/TextInput/useTextInput.js#L42

---

This update is totally not related to this PR's change. However, I thought it may be useful in shedding some light in how we may be able to improve the stability, usability, reliability, and performance for our base components. Hopefully this will come with the adoption of G2 Components, either in it's spirit, it's parts, or it's totality (🙏 )

File: packages/components/src/range-control/test/index.js
Comment (ntsekouras) [inline packages/components/src/range-control/test/index.js:60]: In this test and the below one you missed the `if` in description.
Also I think a better approach to tests is to test what is expected and not what is shouldn't be.

For example I think it should be like: `expect( rangeInput.value ).toBe(.......`

**Review by ntsekouras — COMMENTED** [review]
Great work here @ItsJonQ 💯 Thanks!

I've left some comments but nothing serious :)

File: packages/components/src/range-control/README.md
Comment (ItsJonQ) [reply packages/components/src/range-control/README.md:241]: Oh strange! Maybe something happened with the revert. Will fix. Thank you!

File: packages/components/src/range-control/test/index.js
Comment (ItsJonQ) [reply packages/components/src/range-control/test/index.js:60]: Great suggestion! 🙏 . Will make that update

File: packages/components/src/number-control/test/index.js
Comment (ItsJonQ) [reply packages/components/src/number-control/test/index.js:149]: I think that was from the revert. It wasn't something I changed in this PR.
I believe using `fireKeyDown` was an improvement made by @stokesman .

Personally, I'd like to move away from `react-dom/test-utils` 🙈 
Perhaps in another PR :)

File: packages/components/src/range-control/test/index.js
Comment (ItsJonQ) [reply packages/components/src/range-control/test/index.js]: Thank you! Fixed :D

File: packages/components/src/number-control/test/index.js
Comment (ItsJonQ) [reply packages/components/src/number-control/test/index.js:91]: Strangely enough, `input[type="number"]` return string values (I think it's a bit strange).
I made `NumberControl` preserve that 😄

**Comment by ItsJonQ** [comment]
@ntsekouras Thank you so much! I applied the changes.

Once tests are green and happy, I can merge it up!

**Review by ntsekouras — APPROVED** [review]
Looks good! Green CI and land this 💯

**Comment by ItsJonQ** [comment]
Something's up with the E2E tests.

Previously, Admin 2 failed (due to a timeout).
I ran the tests again.
Now Admin 2 is successful, but Admin 3 failed.

I'm going to re-run them one more time 🤞

**Comment by stokesman** [comment]
I'd like it if we can hold off merging this for just a bit for two reasons. I did retry the PR yesterday after 68aa804 and still found some issues (that may not need to considered in-scope of this PR). But the larger reason is I've tinkered a bit and found some non-reverting changes that fix this (and offer a small unlocked improvement to `UnitControl`) just need to clean up and put it in a PR.

**Comment by stokesman** [comment]
I've updated my previous PR #25933. It could do with maybe a little more description or screen recordings but the changes are pretty mild.

@ItsJonQ, I don't know, but I think you might like the `UnitControl` update there even if we decide to go with this PR.

**Comment by ItsJonQ** [comment]
@stokesman Interesting! I took a quick look at your PR. I didn't dive in too closely as it was still a draft PR. I'd be happy to give it a spin if you think it's ready!

> I did retry the PR yesterday after 68aa804 and still found some issues (that may not need to considered in-scope of this PR)

Oh dear! Do you mind sharing what those issues may be? I'd mostly like to know improve my awareness of edge/test cases 😊

**Comment by stokesman** [comment]
@ItsJonQ, I think besides a comment I forgot to finish updating or other nitpicks that PR is good to go.

I was also meaning to update you on the issues I mentioned which was actually just the previous one I brought up wasn't fully solved. That is the number input of the range control for the Spacer block could still fail to update if the in-canvas resizer was dragged once the input had been cleared. It's so obscure and I think existed before #24460 so I'd not insist it matters for this PR.

The only other thing I can think of at the moment is tangential because it also existed before #24460 and it's more to do with the reset behavior and may be the responsibility of consumers to fix in their implementations. I can add some more detail in case you'd like.

**Comment by ItsJonQ** [comment]
Update: I tested out the PR from @stokesman :
https://github.com/WordPress/gutenberg/pull/25933

It looks good! We can proceed with that one instead of this revert PR.
We can revisit things if issues persist :)

**Comment by ItsJonQ** [comment]
Closing this in favour of https://github.com/WordPress/gutenberg/pull/25933

**Pull request closed without merging**
